src/board/system76/common/power.c

```diff
@@ -334,13 +334,42 @@ void power_off(void) {
     update_power_state();
 }
 
-// Set the CPU power limit appropriately
-static bool power_peci_limit(bool ac) {
+bool power_peci_limit(bool ac) {
     if (peci_available()) {
-        uint16_t watts = ac ? POWER_LIMIT_AC : POWER_LIMIT_DC;
-        // Set PL4 using PECI
-        int16_t res = peci_wr_pkg_config(60, 0, ((uint32_t)watts) * 8);
-        DEBUG("power_peci_limit %d = %d\n", watts, res);
+        uint32_t watts;
+
+        if (ac) {
+            // AC adapter rating with a 15% margin for conversion  losses & spikes
+            // This results in average pwr consumption ~equal to adapter rating
+            // TODO handle this differently for models with HPB enabled
+            watts = (uint32_t)battery_charger_input_current *
+                (uint32_t)battery_charger_input_voltage * 85 / 100000;
+        } else {
+            watts = POWER_LIMIT_DC;
+        }
+
+        uint8_t retry = 10;
+        uint16_t res;
+        do {
+            res = peci_wr_pkg_config(PECI_REG_PKG_CFG_PSYS_PL2, 0, PECI_PSYS_PL2(watts));
+            delay_us(1);
+        } while (res != 0x40 && retry--);
+        DEBUG("PECI SET PsysPL2 %ld = %d\n", watts, res);
+
+        retry = 10;
+        do {
+            res = peci_wr_pkg_config(PECI_REG_PKG_CFG_PL3, 0, PECI_PL3(watts, 6, 4));
+            delay_us(1);
+        } while (res != 0x40 && retry--);
+        DEBUG("PECI SET PL3 %ld = %d\n", watts, res);
+
+        retry = 10;
+        do {
+            res = peci_wr_pkg_config(PECI_REG_PKG_CFG_PL4, 0, PECI_PL4(watts));
```

Comment:
Not sure if PL4 should be the same as PL2/PL3. PL4 is Potential Pewak Power of the SoC, but what about the ROP (Rest of Platform)? I think it should be slightly lower than psysPL2 (say 5W or 10W at least)

Reply:
To be honest I duplicated chromebook behavior here. But yes, PL4 can be set lower, perhaps to `POWER_LIMIT_AC` when on AC power.

Reply:
> ROP is included in the Psys measurements and is handled by Psys limits.

True, but the thing is PL4 does not account for ROP.

Reply:
Okay, set PL4 to POWER_LIMIT_AC, when power supply capacity allows.